### ISSUE TITLE
AttributeError: module 'lib' has no attribute 'X509_V_FLAG_NOTIFY_POLICY'. Did you mean: 'X509_V_FLAG_EXPLICIT_POLICY'?

### ISSUE THREAD

**Issue by mxmlnkn** [ronf/asyncssh#690]
`python3 --version` : `Python 3.10.12`

Installation:

```
pip install --force-reinstall asyncssh
```

Output:

```
pip install --force-reinstall asyncssh
Defaulting to user installation because normal site-packages is not writeable
Collecting asyncssh
  Using cached asyncssh-2.17.0-py3-none-any.whl.metadata (10.0 kB)
Collecting cryptography>=39.0 (from asyncssh)
  Using cached cryptography-43.0.1-cp39-abi3-manylinux_2_28_x86_64.whl.metadata (5.4 kB)
Collecting typing-extensions>=4.0.0 (from asyncssh)
  Using cached typing_extensions-4.12.2-py3-none-any.whl.metadata (3.0 kB)
Collecting cffi>=1.12 (from cryptography>=39.0->asyncssh)
  Using cached cffi-1.17.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (1.5 kB)
Collecting pycparser (from cffi>=1.12->cryptography>=39.0->asyncssh)
  Using cached pycparser-2.22-py3-none-any.whl.metadata (943 bytes)
Using cached asyncssh-2.17.0-py3-none-any.whl (363 kB)
Using cached cryptography-43.0.1-cp39-abi3-manylinux_2_28_x86_64.whl (4.0 MB)
Using cached typing_extensions-4.12.2-py3-none-any.whl (37 kB)
Using cached cffi-1.17.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (446 kB)
Using cached pycparser-2.22-py3-none-any.whl (117 kB)
Installing collected packages: typing-extensions, pycparser, cffi, cryptography, asyncssh
  Attempting uninstall: typing-extensions
    Found existing installation: typing_extensions 4.12.2
    Uninstalling typing_extensions-4.12.2:
      Successfully uninstalled typing_extensions-4.12.2
  Attempting uninstall: pycparser
    Found existing installation: pycparser 2.22
    Uninstalling pycparser-2.22:
      Successfully uninstalled pycparser-2.22
  Attempting uninstall: cffi
    Found existing installation: cffi 1.17.1
    Uninstalling cffi-1.17.1:
      Successfully uninstalled cffi-1.17.1
  Attempting uninstall: cryptography
    Found existing installation: cryptography 43.0.1
    Uninstalling cryptography-43.0.1:
      Successfully uninstalled cryptography-43.0.1
  Attempting uninstall: asyncssh
    Found existing installation: asyncssh 2.17.0
    Uninstalling asyncssh-2.17.0:
      Successfully uninstalled asyncssh-2.17.0
Successfully installed asyncssh-2.17.0 cffi-1.17.1 cryptography-43.0.1 pycparser-2.22 typing-extensions-4.12.2
```

Trying to import the module with `python3 -c 'import asyncssh'` yields:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/user/.local/lib/python3.10/site-packages/asyncssh/__init__.py", line 31, in <module>
    from .agent import SSHAgentClient, SSHAgentKeyPair, connect_agent
  File "/home/user/.local/lib/python3.10/site-packages/asyncssh/agent.py", line 34, in <module>
    from .public_key import KeyPairListArg, SSHCertificate, SSHKeyPair
  File "/home/user/.local/lib/python3.10/site-packages/asyncssh/public_key.py", line 37, in <module>
    from .crypto import ed25519_available, ed448_available
  File "/home/user/.local/lib/python3.10/site-packages/asyncssh/crypto/__init__.py", line 58, in <module>
    from .x509 import X509Certificate, X509Name, X509NamePattern
  File "/home/user/.local/lib/python3.10/site-packages/asyncssh/crypto/x509.py", line 32, in <module>
    from OpenSSL import crypto
  File "/usr/lib/python3/dist-packages/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import crypto, SSL
  File "/usr/lib/python3/dist-packages/OpenSSL/crypto.py", line 1579, in <module>
    class X509StoreFlags(object):
  File "/usr/lib/python3/dist-packages/OpenSSL/crypto.py", line 1598, in X509StoreFlags
    NOTIFY_POLICY = _lib.X509_V_FLAG_NOTIFY_POLICY
AttributeError: module 'lib' has no attribute 'X509_V_FLAG_NOTIFY_POLICY'. Did you mean: 'X509_V_FLAG_EXPLICIT_POLICY'?
```

I found [this](https://github.com/oobabooga/text-generation-webui/issues/5914#issuecomment-2086842883) user with the same problem and some workaround, i.e., upgrading `pyopenssl`. I am using Ubuntu 22.04.4 LTS, which ships with:
```
dpkg -l '*python3-openssl*'
||/ Name                Version      Architecture Description
+++-===================-============-============-===========================================
ii  python3-openssl     21.0.0-1     all          Python 3 wrapper around the OpenSSL library
```

PyPI [lists](https://pypi.org/project/pyOpenSSL/) version 24.2.1 as the most recent, so 3 major versions further.
I tried upgrading version by version. I get the same error with pyopenssl 22, but with version 23 it starts to work:

```
python3 -m pip install 'pyopenssl>=23'
```

```
Defaulting to user installation because normal site-packages is not writeable
Collecting pyopenssl>=23
  Using cached pyOpenSSL-24.2.1-py3-none-any.whl.metadata (13 kB)
Collecting cryptography<44,>=41.0.5 (from pyopenssl>=23)
  Using cached cryptography-43.0.1-cp39-abi3-manylinux_2_28_x86_64.whl.metadata (5.4 kB)
Requirement already satisfied: cffi>=1.12 in /home/user/.local/lib/python3.10/site-packages (from cryptography<44,>=41.0.5->pyopenssl>=23) (1.17.1)
Requirement already satisfied: pycparser in /home/user/.local/lib/python3.10/site-packages (from cffi>=1.12->cryptography<44,>=41.0.5->pyopenssl>=23) (2.22)
Using cached pyOpenSSL-24.2.1-py3-none-any.whl (58 kB)
Using cached cryptography-43.0.1-cp39-abi3-manylinux_2_28_x86_64.whl (4.0 MB)
Installing collected packages: cryptography, pyopenssl
  Attempting uninstall: cryptography
    Found existing installation: cryptography 39.0.2
    Uninstalling cryptography-39.0.2:
      Successfully uninstalled cryptography-39.0.2
Successfully installed cryptography-43.0.1 pyopenssl-24.2.1
```

The import now works fine:

```
python3 -c 'import asyncssh; print(asyncssh.__version__)'  # 2.17.0
```

I think pyopenssl should be pinned to >= 23 to avoid this issue.

**Comment by ronf** [comment]
The current version of AsyncSSH already pins pyOpenSSL to >= 23.0.0, and it has actually been this way for about a year now, dating back to AsyncSSH 2.14.0. So, I'm not sure why you ran into this error.

**Comment by ronf** [comment]
Perhaps the issue is that pyOpenSSL is an optional package, and so your install didn't request it, but you had an older version installed. AsyncSSH will run without pyOpenSSL, but it can't necessarily handle it being installed but outdated. If you specify pyOpenSSL as part of the install, it should pick up the right version, but if you install AsyncSSH without that option, it won't touch what's there.

**Comment by mxmlnkn** [comment]
Yeah, that seems to be exactly the issue, it being only an optional dependency.  But if it is an optional dependency, then there must be some kind of "if module is installed" check and that check could be extended to also check the major version.

**Comment by ronf** [comment]
I'm not sure there's a way to check the major version until after it is imported, and from what you've said the error you saw happened in the import. If that's the case, though, perhaps I could catch an exception on the import. I'll experiment with it.

In the meantime, running `pip install 'asyncssh[pyOpenSSL]' should work.

**Comment by ronf** [comment]
Does the following change in `crypto/__init__.py` work for you?

```patch
diff --git a/asyncssh/crypto/__init__.py b/asyncssh/crypto/__init__.py
index ef6c8e5..3515a84 100644
--- a/asyncssh/crypto/__init__.py
+++ b/asyncssh/crypto/__init__.py
@@ -57,5 +57,5 @@ except (ImportError, AttributeError, OSError): # pragma: no cover
 try:
     from .x509 import X509Certificate, X509Name, X509NamePattern
     from .x509 import generate_x509_certificate, import_x509_certificate
-except ImportError: # pragma: no cover
+except (AttributeError, ImportError): # pragma: no cover
```

With this change, importing asyncssh succeeds, but it sets a flag internally to prevent X.509 operations. For example:

```
>>> k = asyncssh.generate_private_key('ssh-rsa')
>>> k.generate_x509_user_certificate(k, 'CN=test')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/private/tmp/asyncssh/public_key.py", line 890, in generate_x509_user_certificate
    return self._generate_x509_certificate(user_key, subject, issuer,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/asyncssh/public_key.py", line 377, in _generate_x509_certificate
    raise KeyGenerationError('X.509 certificate generation '
asyncssh.public_key.KeyGenerationError: X.509 certificate generation requires PyOpenSSL
```

**Comment by mxmlnkn** [comment]
That error message looks helpful. The proposed patch works for me. (I can reproduce the error by uninstalling pyopenssl, which leaves me with the Ubuntu-managed pyopenssl 21.0.0-1)

**Comment by ronf** [comment]
Great, thanks for the confirmation! This is now available in the "develop" branch as commit 2601966 and will be included in the next release.

**Comment by mxmlnkn** [comment]
Much appreciated.
